### PR TITLE
New ABC sage.structure.element.NumberFieldElement, deprecate is_NumberFieldElement

### LINKED ISSUE #34931: New ABC sage.structure.element.NumberFieldElement, deprecate is_NumberFieldElement
<div id="comment:0"></div>

We also make `sage.rings.number_field` a namespace package. This is necessary because `.number_field_base` is needed in sagemath-categories (#32432), whereas implementation modules depend on NTL.

Part of Meta-ticket #32414

CC:  @tscrim

Component: **refactoring**

Author: **Matthias Koeppe**

_Issue created by migration from https://trac.sagemath.org/ticket/34931_

### PR DESCRIPTION
**Pull request by mkoeppe**
### 📚 Description
Fixes #34931

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### ⌛ Dependencies

Depends on #35033

### CODE REVIEW

File: src/sage/rings/integer_ring.pyx [hunk 1]
```diff
@@ -60,6 +60,7 @@ from sage.categories.basic import EuclideanDomains
 from sage.categories.infinite_enumerated_sets import InfiniteEnumeratedSets
 from sage.structure.coerce cimport is_numpy_type
 from sage.structure.element cimport parent
+from sage.structure.element import NumberFieldElement
```
Comment (tscrim) [inline src/sage/rings/integer_ring.pyx]: Why is this not a `cimport` (and combined with the previous line)?

**Review by tscrim — CHANGES_REQUESTED** [review]

File: src/sage/rings/integer_ring.pyx (hunk 1 above)
Comment (mkoeppe) [reply src/sage/rings/integer_ring.pyx]: I don't think it matters very much for an `isinstance` call

Comment (tscrim) [reply src/sage/rings/integer_ring.pyx]: It might not matter much, but it still is within `integer_ring.pyx`, which we should make sure to optimize every chance we get.

Comment (mkoeppe) [reply src/sage/rings/integer_ring.pyx]: I've made this change.

**Comment by mkoeppe** [comment]
By the way, did you see https://github.com/sagemath/sage/issues/34931#issuecomment-1418193122?

**Review by tscrim — APPROVED** [review]
Thank you. LGTM.

**Comment by tscrim** [comment]
> By the way, did you see [#34931 (comment)](https://github.com/sagemath/sage/issues/34931#issuecomment-1418193122)?

I did, but I don't know which is better either. At this point, I think it is best to keep with the same pattern as it makes it easier to refactor stuff later on if something changes.

**Comment by mkoeppe** [comment]
same pattern as what?

**Comment by tscrim** [comment]
> same pattern as what?

As the other ABCEDFITs in `element.pxd/pyx`.

**Comment by mkoeppe** [comment]
Note that in #35076 (for `CommutativePolynomial` etc.), we moved away from that again...

**Comment by tscrim** [comment]
I forgot that. Hmm...then perhaps we should put it into its own little file. I am worried a bit about bloat within `element.p*`. It does make a bit more sense within `element.pyx` if it was a true ABC that people could subclass of off it. Although since it is not meant for that, it could easily get confused.

**Comment by mkoeppe** [comment]
OK, then I'll make this change.

**Review by tscrim — APPROVED** [review]
Thanks. LGTM.

**Comment by tornaria** [comment]
@mkoeppe : for some reason I don't understand, the removal of `.../number_field/__init__.py` causes `number_field_element_base` to be unimportable:
```python
$ PYTHONPATH=pkgs/sagemath-standard/build/lib.linux-x86_64-cpython-311 python
Python 3.11.2 (main, Feb  8 2023, 14:30:35) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sage.rings.number_field.number_field_element_base
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'sage.rings.number_field.number_field_element_base'
>>> 
$ touch pkgs/sagemath-standard/build/lib.linux-x86_64-cpython-311/sage/rings/number_field/__init__.py
$ PYTHONPATH=pkgs/sagemath-standard/build/lib.linux-x86_64-cpython-311 python
Python 3.11.2 (main, Feb  8 2023, 14:30:35) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sage.rings.number_field.number_field_element_base
>>> 
```
I'm doing it here with plain python, but the same thing happens with sage.

There are lots of empty `__init__.py` files:
```
$ find -name __init__.py  -size 0 | wc -l
327
```

**Comment by tornaria** [comment]
This is what I think happens:
 - I have sage installed system-wide
 - Hence there is an (empty) file `/usr/lib/python3.11/site-packages/sage/rings/number_field/__init__.py`
 - When doing `import sage.rings.number_field.number_field_element_base`, python tries first to find the module `sage.rings.number_field`
 - Since `PYTHONPATH=pkgs/sagemath-standard/build/lib.linux-x86_64-cpython-311`, it looks there for this module. Without the empty `__init__.py` there, this fails.
 - Python still wants `sage.rings.number_field`, so it keeps looking and finds it at the system `site-packages`.
 - Now the submodule `number_field_element_base` is searched _only_ at `site-packages` but my system install is 9.8 so it doesn't contain it, hence the failure.

If I'm correct, this explains why only `number_field_element_base` fails to import. But on the other hand it means that importing other submodules of `number_field` will pick them from the `site-packages` and not from the current location.

In other words, a system install of sagemath will "shadow" a user install; in principle this shouldn't be an issue, except when the user install is missing a module python will pick it from the system install. The only way to avoid this seems to be to keep empty modules to 'hide" the corresponding system modules.

**Comment by tornaria** [comment]
> > Since `sage.rings.number_field` is a package (it contains modules) the `__init__.py` file is required.
> 
> No, this is no longer true since PEP 420. Read https://doc.sagemath.org/html/en/developer/packaging_sage_library.html#ordinary-packages-vs-implicit-namespace-packages

I fail to see why you needed to remove
```
src/sage/rings/number_field/__init__.py
src/sage/rings/polynomial/__init__.py
src/sage/rings/finite_rings/__init__.py
```
You are not moving these modules to a separate package, are you? They are still part of `sagemath-standard` AFAICT.

If kind of silly to modularize sage-the-library breaking it in smaller more manageable parts, if making system packages for those parts is not supported. The whole point of installing sage-the-library in the system site-packages is so that it can be imported from regular python. At least I thought that was the plan...

If  you really need to remove those `__init__.py` files please document what is broken when they are present and/or give me a way out of this mess.

----

These are the directories that may be missing `__init__.py`:
```
$ for d in $(find src/sage -type d) ; do ls $d/__init__* &> /dev/null || ((ls $d/*.py* &> /dev/null || ls $d/*.px* &> /dev/null) && echo $d ) ; done
src/sage
src/sage/matrix
src/sage/ext
src/sage/rings
src/sage/rings/number_field
src/sage/rings/polynomial
src/sage/rings/finite_rings
src/sage/numerical
src/sage/numerical/backends
src/sage/graphs
src/sage/graphs/graph_decompositions
src/sage/sets
src/sage/categories
src/sage/interfaces
src/sage/misc
src/sage/libs
src/sage/arith
src/sage/ext_data/nbconvert
```
The last one is a false positive, but all the others seem reasonable to me. I don't quite care if the packages have or not `__init__.py` files. What causes the problem is changing a package from having to not-having `__init__.py`.

**Comment by mkoeppe** [comment]
> I fail to see why you needed to remove
> 
> ```
> src/sage/rings/number_field/__init__.py
> src/sage/rings/polynomial/__init__.py
> src/sage/rings/finite_rings/__init__.py
> ```
> 
> You are not moving these modules to a separate package, are you?

Yes, I *am* moving some of the contained modules to separate distributions.

Technology preview: https://github.com/sagemath/sage/pull/35095 – this enlarges the distribution *sagemath-categories* to include a bit from each of these 3 namespace packages. See https://github.com/sagemath/sage/blob/2b264b236844b196e360de300701fbe5fc977636/pkgs/sagemath-categories/MANIFEST.in.m4
It includes the base classes of number fields, but no number field implementations other than the rationals.
It includes polynomials, enough for basic operations with univariate and multivariate polynomials, but no specialized element implementations that pull in dependencies on various libraries.
Likewise for finite rings.

**Comment by mkoeppe** [comment]
> They are still part of `sagemath-standard` AFAICT.

Yes, at the moment **sagemath-standard** is still monolithic. But https://github.com/sagemath/sage/issues/34587 makes sagemath-standard depend on subset distributions such as sagemath-categories.

**Comment by tornaria** [comment]
> > You are not moving these modules to a separate package, are you?

> Yes, I _am_ moving some of the contained modules to separate distributions.

Is this happening for 10.0? If not, can we just keep these at least until 10.0 is released? Better yet, can you remove them as part of the actual move to separate distributions? I assume that will be a lot of pain for distributions, we may as well get all the pain at the same time. Maybe I'll get my sagemath package stuck at 10.0 forever.

Another thought: why is it so important to remove `__init__.py` ? Can't we just keep it in sagemath-standard? The same file cannot belong to two separate distributions, but it could belong to the main one, couldn't it?

I still fail to see what problem this is solving.

**Comment by mkoeppe** [comment]
> why is it so important to remove `__init__.py` ? Can't we just keep it in sagemath-standard? The same file cannot belong to two separate distributions, but it could belong to the main one, couldn't it?

No. If any distribution ships an `__init__.py` file, then it makes the portions of that package provided by other distributions invisible.

**Comment by mkoeppe** [comment]
I don't know what's special about 10.0

**Comment by mkoeppe** [comment]
> I still fail to see what problem this is solving.

The modularization solves the problem that nobody can install and use parts of Sage without pulling in gigabytes of non-pip-installable dependencies.

**Comment by mkoeppe** [comment]
I do agree that some synchronization may help – because the issue that you have reported arises from the *transition* to namespace packages; it is not a problem by itself. 
So I will prepare a PR that removes all the `__init__.py` files of all packages that are intended to become namespace packages.

**Comment by tornaria** [comment]
> I don't know what's special about 10.0

Nothing special, I'm just saying don't break releases if the PR that need this are not going to be merged.

> > I still fail to see what problem this is solving.
> The modularization solves the problem that nobody can install and use parts of Sage without pulling in gigabytes of non-pip-installable dependencies.

That's what distros are for. The sagemath package in void linux is just 50M compressed (builds in 10 minutes from 28M source). What is the advantage of shipping it in separate chunks, if they are developed together and they can't be updated separately?
 
> I do agree that some synchronization may help – because the issue that you have reported arises from the _transition_ to namespace packages; it is not a problem by itself. So I will prepare a PR that removes all the `__init__.py` files of all packages that are intended to become namespace packages.

Now, that's constructive. Let's do it explicitly and *knowing* we break stuff. What about this:
a. we restore in the next beta the 3 empty `__init__.py` files that were removed in beta5, so we don't break development in the presence of a system install of 9.8
b. we coordinate with distros so that the system install of sage 10.0 *will not include a given list of empty  `__init__.py` files*
c. after 10.0 is available on distros without these empty files, we remove them in a future release cycle.

Maybe if there was an _easy and well documented_ way to install sage-the-library system wide (e.g. for a distro) we could impose some policy to distro-packages (like not installing the empty `__init__.py` files in site-packages).

----

Let me ask you a question in the opposite direction: do you know why are all those empty `__init__.py` files needed for? IOW, what would happen if I just remove _all_ the empty `__init__.py` files from site-packages? As I understand PEP 420, there is no reason to keep them if python 2 support is not needed. Is there a measurable effect of removing all them?

**Comment by mkoeppe** [comment]
> > > The modularization solves the problem that nobody can install and use parts of Sage without pulling in gigabytes of non-pip-installable dependencies.
> 
> That's what distros are for.

There are lots of people out there who don't use distro Python.

Basically, if something is not pip-installable, it's not usable as an upstream for other Python projects.

**Comment by mkoeppe** [comment]
By the way, does Void have any mechanism for users to make Python venvs that mix a defined set of Void-packaged Python distribution packages with pip-installed Python packages?

**Comment by mkoeppe** [comment]
> Let me ask you a question in the opposite direction: do you know why are all those empty `__init__.py` files needed for? IOW, what would happen if I just remove _all_ the empty `__init__.py` files from site-packages? As I understand PEP 420, there is no reason to keep them if python 2 support is not needed. Is there a measurable effect of removing all them?

There is no technical need for any of the empty `__init__.py` files. You can remove them.
Basically, they are in our source tree because they have not been removed. However:
- In our build system, when there is no `__init__.py` file, there needs to be an `all*.py` file to ensure that the modules inside it are found.
- It _may_ be valuable to keep some `__init__.py` files as documentation that this package is not intended to be made a namespace packages, as in "it's not welcome for other distributions to add stuff here".
